packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart

```diff
@@ -410,417 +565,363 @@ Uri nativeAssetsBuildUri(Uri projectUri, OS os) {
-        projectUri,
-        buildRunner,
-      ),
-  ];
-  final Uri nativeAssetsUri = await writeNativeAssetsYaml(
-    KernelAssets(nativeAssetPaths),
-    buildUri,
-    fileSystem,
-  );
-  return nativeAssetsUri;
-}
-
-/// With `flutter run -d all` we need a place to store the native assets
-/// mapping for multiple OSes combined.
-Uri buildUriMultiple(Uri projectUri) {
-  final String buildDir = build_info.getBuildDirectory();
-  return projectUri.resolve('$buildDir/native_assets/multiple/');
 }
 
-/// Dry run the native builds.
+/// Invokes the build of all transitive Dart packages.
 ///
-/// This does not build native assets, it only simulates what the final paths
-/// of all assets will be so that this can be embedded in the kernel file.
-Future<Uri?> dryRunNativeAssetsSingleArchitecture({
-  required NativeAssetsBuildRunner buildRunner,
+/// This will invoke `hook/build.dart` and `hook/link.dart` (if applicable) for
+/// all transitive dart packages that define such hooks.
+Future<DartBuildResult> _runDartBuild({
+  required Map<String, String> environmentDefines,
+  required FlutterNativeAssetsBuildRunner buildRunner,
+  required List<Target> targets,
   required Uri projectUri,
-  bool flutterTester = false,
+  required BuildModeImpl buildMode,
   required FileSystem fileSystem,
-  required OSImpl os,
+  required OS targetOS,
 }) async {
-  if (!await nativeBuildRequired(buildRunner)) {
-    return null;
+  final bool linkingEnabled = buildMode == BuildMode.release;
+  final String targetString = targets.length == 1
+      ? targets.single.toString()
+      : targets.toList().toString();
+  globals.logger
+      .printTrace('Building native assets for $targetString $buildMode.');
+  final List<Asset> assets = <Asset>[];
+  final Set<Uri> dependencies = <Uri>{};
+  final build_info.EnvironmentType? environmentType;
+  if (targetOS == OS.iOS) {
+    final String? sdkRoot = environmentDefines[build_info.kSdkRoot];
+    if (sdkRoot == null) {
+      throw MissingDefineException(build_info.kSdkRoot, 'native_assets');
+    }
+    environmentType = xcode.environmentTypeFromSdkroot(sdkRoot, fileSystem);
+  } else {
+    environmentType = null;
   }
 
-  final Uri buildUri = nativeAssetsBuildUri(projectUri, os);
-  final Iterable<KernelAsset> nativeAssetPaths = await dryRunNativeAssetsSingleArchitectureInternal(
-    fileSystem,
-    projectUri,
-    flutterTester,
-    buildRunner,
-    os,
-  );
-  final Uri nativeAssetsUri = await writeNativeAssetsYaml(
-    KernelAssets(nativeAssetPaths.toList()),
-    buildUri,
-    fileSystem,
-  );
-  return nativeAssetsUri;
-}
+  final CCompilerConfigImpl cCompilerConfig = targetOS == OS.android
+      ? await buildRunner.ndkCCompilerConfigImpl
+      : await buildRunner.cCompilerConfig;
+  final IOSSdkImpl? targetIOSSdk =
+      targetOS == OS.iOS ? getIOSSdk(environmentType!) : null;
+  // TODO(dcharkes): Fetch minimum iOS version from somewhere. https://github.com/flutter/flutter/issues/145104
```

Comment:
OS-specific things should live in the OS-specific files. Maybe move helper functions there (also for targetIOSVersion and targetMacOSVersion) and leave a comment on the `buildRunner.build` invocation it contains the config for all OSes. And then on the helper functions that they return null for the irrelevant OSes?

Reply:
Done.